### ISSUE TITLE
Error in chronos function

### ISSUE THREAD

**Issue by takaW496** [emmanuelparadis/ape#13]
Thank you for the development and maintenance of ape. We use chronos function in our analysis pipeline and it functions nicely in almost all cases. However, I got an error in a handful cases as below. I confirmed this error in ape version 5.5 and 5.4_1. The R image file is attached for reproducibility. We could not figure out what the cause is. I would appreciate if you could take a look and give me any suggestions. Thank you.

[error_test.RData.zip](https://github.com/emmanuelparadis/ape/files/6458063/error_test.RData.zip)

```
chronos_out = ape::chronos(gn_tree, lambda = chronos_lambda, model = chronos_model, calibration = current_calibration_table, control = chronos_control)

Setting initial dates...
Fitting in progress... get a first set of estimates
Error in if (sum(freq) > 1) return(-1e+100) : 
  missing value where TRUE/FALSE needed
```

**Comment by emmanuelparadis** [comment]
Thanks for the appreciation!
Your example works using the default model:

```r
chronos_out <- chronos(gn_tree, lambda = chronos_lambda, calibration = current_calibration_table, control = chronos_control)
```
or using a smaller number of categories of rates:
```r
chronos_control$nb.rate.cat <- 1
chronos_out <- chronos(gn_tree, lambda = chronos_lambda, model = chronos_model, calibration = current_calibration_table, control = chronos_control)
```
But 2 categories gives the same error than with 10.
The relaxed (gamma distributed rates) model works also.
Anyway, the discrete model with 2 categories should work: there is maybe a bug. I'll have a close look at it.

**Comment by takaW496** [comment]
Thank you for checking the error. I look forward to your updates.

**Comment by kfuku52** [comment]
Hi, I've been a big fan of ape for almost a decade and would like to thank you for your long-term maintenance and development of this excellent package. I'm working together with @takaW496 and got the same error in some of my datasets too. This error seems to come from NaNs in the vectors `freq` and `read.edge.length` produced internally in `chronos()`. When monitored with @takaW496 's data, these vectors starts as
```
[1] "freq"
[1] 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1 0.1
[1] "real.edge.length"
 [1] 3.560598e+01 3.560598e+01 3.560598e+01 3.560598e+01 3.560598e+01
 [6] 1.017314e+01 1.017314e+01 1.017314e+01 1.017314e+01 1.017314e+01
[11] 1.017314e+01 1.017314e+01 1.017314e+01 2.034627e+01 1.525971e+01
[16] 1.525971e+01 1.525971e+01 4.069255e+01 1.695523e+01 1.695523e+01
[21] 1.695523e+01 8.477614e+00 8.477614e+00 8.477614e+00 1.695523e+01
[26] 1.695523e+01 8.477614e+00 8.477614e+00 4.238807e+00 4.238807e+00
[31] 1.412936e+00 1.412936e+00 1.412936e+00 7.064679e-01 7.064679e-01
[36] 3.532339e-01 3.532339e-01 1.177446e-01 1.177446e-01 1.177446e-01
[41] 1.177446e-01 1.177446e-01 1.177446e-01 5.887232e-02 5.887232e-02
[46] 2.943616e-02 2.943616e-02 9.812054e-03 9.812054e-03 9.812054e-03
[51] 9.812054e-03 9.812054e-03 9.812054e-03 4.906027e-03 4.906027e-03
[56] 2.453013e-03 2.453013e-03 6.132533e-04 6.132533e-04 6.132533e-04
[61] 6.132533e-04 1.226507e-04 1.226507e-04 1.226507e-04 1.226507e-04
[66] 1.226507e-04 1.226507e-04 2.453013e-04 3.679520e-04 4.906027e-04
[71] 1.226507e-03 1.839760e-03 2.825871e+00 1.220776e+01 1.220776e+01
[76] 1.220776e+01 1.220776e+01 1.220776e+01 1.220776e+01 2.441553e+01
[81] 3.662329e+01 4.883106e+01 1.068179e+02 1.424239e+02
```
And the stored values were completely identical over 61 rounds of the `nlminb()` optimization, but suddenly in 62nd round, NaNs are produced like this.
```
[1] "freq"
[1] NaN NaN NaN NaN NaN NaN NaN NaN NaN
[1] "real.edge.length"
 [1]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
 [9]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[17]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[25]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[33]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[41]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[49]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[57]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[65]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[73]      NaN      NaN      NaN      NaN      NaN      NaN      NaN      NaN
[81]      NaN      NaN      NaN 142.4239
```
NaNs seem to be produced at a low frequency, probably once in dozens or hundreads of trials. I feel that this is a bug or unexpected behavior of `nlminb()`, although I'm not very familiar with this function. To avoid this error anyway, we can simply check the presence of NaNs and `return` earlier if detected. The returned placeholder loglik should also be changed from `-1e+100` to `-Inf`, otherwise `nlminb()` can interpret this is the best set of parameters and `chronos_out$edge.length` ends up all NaNs. `chronos` successfully completed after applying such change, which you can see in this commit: https://github.com/kfuku52/ape-1/commit/6c7461f4a0276336e990ce82ffe073d83e8ef9ba

The problem is not completely solved, because, looking at the log-Lik, the optimization didn't happen at all.
```
Setting initial dates...
Fitting in progress... get a first set of estimates
         (Penalised) log-lik = -Inf 
Optimising rates...
Warning message in nlminb(current.rates, f.rates, g.rates, control = list(eval.max = 1000, :
“NA/NaN function evaluation”
 frequencies... dates... -Inf 

log-Lik = -Inf 
PHIIC = Inf 
is.ultrametric(chronos_out) = TRUE 
```
In conclusion, the `nlminb()` optimization is challenging in this dataset but a not-so-optimized dated tree is obtainable with the above changes. If it looks OK, I'll be happy to create a PR. 

P.S. The `-1e+100`-to-`-Inf` replacement could be applicable to the other part of `chronos` to avoid similar bugs in the future. I also noticed `e1` and `e2` are not passed properly as arguments to some functions.

![image](https://user-images.githubusercontent.com/19270515/120449513-c05d1f00-c38f-11eb-8a1e-d9dffebfa6cd.png)

**Comment by emmanuelparadis** [comment]
Hi Kenji. Thanks for the detailed report. I'll have a close look at it. I guess you are right that just a simple check might be enough. I'll get back to you soon.
And thanks a lot for the appreciaiton ;)

**Comment by emmanuelparadis** [comment]
I've fixed this (I hope!) The problem came from the fact that the branch lengths are extremely heterogeneous, some being very large, and these are used to set initial values for the rates by first estimating a rough rate for each branch, then calculating regular intervals from the smallest to the largest values of these rates. Because the initial rates are too heterogeneous, this resulted in some likelihood values zero for all rate categories for some branches, and then finally in infinite likelihoods. Quantiles are now used instead of regular intervals, giving better results (Kenji: you were right. It's a challenging data set!)

I've pushed the fix here. Of course, please give feedback if something's wrong.

Best,
Emmanuel

**Comment by kfuku52** [comment]
Thank you so much! We will report back if there are any other problems.
